src/vs/workbench/contrib/debug/common/debugModel.ts

```diff
@@ -1197,24 +1197,46 @@ export class DataBreakpoint extends BaseBreakpoint implements IDataBreakpoint {
 		this.accessTypes = opts.accessTypes;
 		this.accessType = opts.accessType;
 		if (opts.initialSessionData) {
-			this.sessionDataIdForAddr.set(opts.initialSessionData.session, opts.initialSessionData.dataId);
+			this.sessionDataId.set(opts.initialSessionData.session, opts.initialSessionData.dataId);
 		}
 	}
 
 	async toDAP(session: IDebugSession): Promise<DebugProtocol.DataBreakpoint | undefined> {
-		let dataId: string;
-		if (this.src.type === DataBreakpointSetType.Variable) {
-			dataId = this.src.dataId;
-		} else {
-			let sessionDataId = this.sessionDataIdForAddr.get(session);
-			if (!sessionDataId) {
-				sessionDataId = (await session.dataBytesBreakpointInfo(this.src.address, this.src.bytes))?.dataId;
-				if (!sessionDataId) {
-					return undefined;
+		let dataId = this.sessionDataId.get(session);
```

Comment:
I think we might need to revisit this pattern for breakpoints with `canPersist`. It _kind of_ works for address breakpoints (minus ASLR shenanigans) but what should actually happen is that the value of `canPersist` gets derived from the data `DataBreakpointInfoResponse` and then we persist the data ID.

This breaks down more when we have sources set using frames and variable IDs. When a new session is started, or when VS Code is restarted and the breakpoints are read from storage, those frames and variable references will no longer have any meaning. We'll request them again here in `toDap` but that's likely to fail or point to the wrong addresses. Instead, if we've already translated them to a data ID with `canPersist`, we should just set them with that data ID instead.

I think we still want to keep the `src` in here for display purposes (and we could potentially present a "refresh" action for certain non-`canPersist` breakpoints within a VS Code session) but the resolution of the data ID should be done and kept at the instant the breakpoint it set.

This is a bit egg on my face since I started the pattern of requesting `dataBytesBreakpointInfo` in this method which was misleading 🙈

Reply:
I had a deeper look now and if I understand you correctly, you argue that the resolution of the data breakpoints should be done as early as possible (i.e., as soon as they are added. And what we are actually interested in persisting is only the `dataId` which needs to be valid across multiple sessions anyway. Is that right? I tried to incorporate that in my next commit but skipped the refreshing of breakpoints within a session for now but that would be surely interesting as well.